### ISSUE TITLE
toupper() is inconsistent with gU

### ISSUE THREAD

**Issue by fenuks** [vim/vim#5573]
**Describe the bug**
It is stated in documentation that `toupper()` works `just like applying gU to the string`. It is not in case of German letter `ß`. `toupper()` and `~` leave it as it is, while `gU` does capitalize it.

**To Reproduce**
1. Paste `ß` into buffer
2. Try to capitalize it with `~` key (letter is unchanged)
3. Try `gU` on the letter (letter is capitalized to `SS`)
4. `:echo toupper('ß')` (letter is unchanged)

**Expected behavior**
Expected `toupper()` and `~` to work consistently with `gU` which alters text, as written in documentation.

**Environment (please complete the following information):**
 - Vim version
```
VIM - Vi IMproved 8.2 (2019 Dec 12, compiled Jan 26 2020 11:45:38)
Included patches: 1-148
Compiled by Arch Linux
Huge version without GUI.  Features included (+) or not (-):
+acl               -farsi             -mouse_sysmouse    -tag_old_static
+arabic            +file_in_path      +mouse_urxvt       -tag_any_white
+autocmd           +find_in_path      +mouse_xterm       +tcl/dyn
+autochdir         +float             +multi_byte        +termguicolors
-autoservername    +folding           +multi_lang        +terminal
-balloon_eval      -footer            -mzscheme          +terminfo
+balloon_eval_term +fork()            +netbeans_intg     +termresponse
-browse            +gettext           +num64             +textobjects
++builtin_terms    -hangul_input      +packages          +textprop
+byte_offset       +iconv             +path_extra        +timers
+channel           +insert_expand     +perl/dyn          +title
+cindent           +job               +persistent_undo   -toolbar
-clientserver      +jumplist          +popupwin          +user_commands
-clipboard         +keymap            +postscript        +vartabs
+cmdline_compl     +lambda            +printer           +vertsplit
+cmdline_hist      +langmap           +profile           +virtualedit
+cmdline_info      +libcall           +python/dyn        +visual
+comments          +linebreak         +python3/dyn       +visualextra
+conceal           +lispindent        +quickfix          +viminfo
+cryptv            +listcmds          +reltime           +vreplace
+cscope            +localmap          +rightleft         +wildignore
+cursorbind        +lua/dyn           +ruby/dyn          +wildmenu
+cursorshape       +menu              +scrollbind        +windows
+dialog_con        +mksession         +signs             +writebackup
+diff              +modify_fname      +smartindent       -X11
+digraphs          +mouse             -sound             -xfontset
-dnd               -mouseshape        +spell             -xim
-ebcdic            +mouse_dec         +startuptime       -xpm
+emacs_tags        +mouse_gpm         +statusline        -xsmp
+eval              -mouse_jsbterm     -sun_workshop      -xterm_clipboard
+ex_extra          +mouse_netterm     +syntax            -xterm_save
+extra_search      +mouse_sgr         +tag_binary        
   system vimrc file: "/etc/vimrc"
     user vimrc file: "$HOME/.vimrc"
 2nd user vimrc file: "~/.vim/vimrc"
      user exrc file: "$HOME/.exrc"
       defaults file: "$VIMRUNTIME/defaults.vim"
  fall-back for $VIM: "/usr/share/vim"
Compilation: gcc -c -I. -Iproto -DHAVE_CONFIG_H     -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1       
Linking: gcc   -L. -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -fstack-protector-strong -rdynamic -Wl,-export-dynamic -Wl,-E -Wl,-rpath,/usr/lib/perl5/5.30/core_perl/CORE  -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -L/usr/local/lib -Wl,--as-needed -o vim        -lm -ltinfo -lelf -lnsl    -lacl -lattr -lgpm -ldl   -Wl,-E -Wl,-rpath,/usr/lib/perl5/5.30/core_perl/CORE -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -fstack-protector-strong -L/usr/local/lib  -L/usr/lib/perl5/5.30/core_perl/CORE -lperl -lpthread -ldl -lm -lcrypt -lutil -lc   -L/usr/lib -ltclstub8.6 -ldl -lz -lpthread -lm
```
 - OS: Arch Linux
 - Terminal: Yakuake 3.0.5+

**Additional context**
Wikipedia has [article on capital version of letter](https://en.wikipedia.org/wiki/Capital_%E1%BA%9E), and recently `ẞ` was officially accepted to replace `SS` ligature, I've checked and vim works correctly with it (all operations to lowercase it, yield lowercase `ß` letter). Possibly, it would to better if `gU` and other uppercase operations returned it, as that would make the operation reversible, running `gu` afterwards would revert to the same initial letter.

**Comment by chrisbra** [comment]
possible related: patch [7.3.536](https://github.com/vim/vim/releases/tag/v7.3.536) and the relevant thread: https://groups.google.com/d/msg/vim_dev/POKD8hCbovI/k1ZN9cbkHukJ and there is some special case handling of the `ß` for the `gU` command in [swapchars](https://github.com/vim/vim/blob/a5edb670dcdeddc8520e012430fb256eb8470f6c/src/ops.c#L1434-L1439)

**Comment by dpelle** [comment]
> Possibly, it would to better if gU and other uppercase operations
> returned it, as that would make the operation reversible

ß→SS is not reversible, as not all ss in German can be replaced
with ß when transforming into lowercase.

**Comment by fenuks** [comment]
> ß→SS is not reversible, as not all ss in German can be replaced
with ß when transforming into lowercase.

I meant capital sharp s character – `ẞ`. It is reversible, and vim already correctly transforms it to lowercase `ß`.

**Comment by chrisbra** [comment]
The reason that `toupper()` for `ß` does not work is, that the [Unicode Database](http://unicode.org/Public/UNIDATA/UnicodeData.txt) does not define any uppercase variant for the `U+00DF (LATIN SMALL LETTER SHARP S)`, so Vim won't use `U+1E9E (LATIN CAPITAL LETTER SHARP S)` as being the upper case character for `U+00DF (LATIN SMALL LETTER SHARP S)` when using `toupper()`

However, `U+1E9E` defines `U+00DF` as being the corresponding lower case chracter. I don't know why it is this way, but I will not touch those unicode tables, they are mostly created automatically from the Unicode releases.

for the other issue using `gU` or `~` on lower case sharp s will be fixed in a few moments, once I merge #14018
